server/src/instant/util/storage.clj

```diff
@@ -0,0 +1,207 @@
+(ns instant.util.storage
+  (:require [clojure.string :as string]
+            [instant.util.exception :as ex]
+            [instant.model.app-user :as app-user-model]
+            [instant.model.rule :as rule-model]
+            [instant.model.app-user-refresh-token :as app-user-refresh-token-model]
+            [instant.db.cel :as cel]
+            [instant.util.json :refer [->json <-json]]
+            [instant.storage.s3 :as s3-util]
+            [instant.storage.beta :as storage-beta])
+
+  (:import
+   (java.util UUID)))
+
+
+;; scopes filename to app-id directory
+(defn ->object-key [app-id filename]
+  (str app-id "/" filename))
+
+;; extracts filename from app-id directory scope
+(defn object-key->filename [object-key]
+  (string/join "/" (rest (string/split object-key #"/"))))
+
+(comment
+  (def rules {:code
+              {"$files" {"bind" ["isLoggedIn" "auth.id != null"]
+                         "allow" {"view" "isLoggedIn"
+                                  "create" "data.path.contains('demo')"}}}})
+  (def program (rule-model/get-program! rules "$files" "create"))
+  (cel/eval-program! program {"auth" (cel/->cel-map {"id" "alex"})
+                              "data" (cel/->cel-map {"path" "demo/image.png"})}))
+
+
+(defn assert-storage-permission! [action {:keys [app-id refresh-token filepath rules-override]}]
+  (let [rules (if rules-override
+                rules-override
+                (rule-model/get-by-app-id {:app-id app-id}))
+        current-user (when refresh-token
+                       (app-user-model/get-by-refresh-token {:app-id app-id
+                                                             :refresh-token refresh-token}))
+
+        program (rule-model/get-program! rules "$files" action)]
+    (ex/assert-permitted!
+     :has-storage-permission?
+     ["$files" action]
+     (if-not program
+       ;; deny access by default if no permissions are currently set
+       false
+       ;; otherwise, evaluate the permissions code
+       (cel/eval-program! program
+                          {"auth" (cel/->cel-map (<-json (->json current-user)))
+                           "data" (cel/->cel-map {"path" filepath})})))))
+
+
+(defn upload-image-to-s3 [app-id filename image-url]
+  (let [object-key (->object-key app-id filename)]
+    (storage-beta/assert-storage-enabled! app-id)
+    (s3-util/upload-image-to-s3 object-key image-url)))
+
+(defn create-signed-download-url!
```

Comment:
lmk if the multi-arity situation here is getting a bit too wonky!

basically wanted to support:
- browser client use case (refresh token needed)
- admin/dashboard use case (no refresh token needed to verify permissions)
- testing use case (mock rules)

is this the best way to handle?

Reply:
(cc @stopachka @dwwoelfel)

Reply:
I would just make different functions? 

Here's my concern with the mutli-arity--the function you call during a test when you don't need permissions is the same function you call when you do need permissions. It seems like we're teetering on the edge of a security issue when someone accidentally calls a function with the wrong airty.

If you have different functions with different names, then it will be much more obvious when you do that.

Reply:
sounds good!

Reply:
(realized this was a bit silly anyway, since what I was really testing was the `assert-storage-permission!` function... removed and just added tests for `assert-storage-permission!` below instead)